feat(static-json): implement fs-first static JSON loading with unified timestamped fetch fallback

fzoonoise opened the pull request:
Add `readStaticJson()` utility to support priority-based static JSON access via filesystem, with fallback to HTTPS fetch and timestamp query param. Refactor all existing fetch logic to use this utility and unify config constants.

## Additions  
- Introduced `readStaticJson()` in `utils/read-static-json.ts` to:
  - Prioritize reading static JSON from GCS FUSE-mounted `/statics` directory
  - Fallback to HTTPS fetch with cache-busting timestamp param
- Added new constants (`STATIC_JSON_*`) for relative JSON paths
- Added `API_STORY_GQL_ENDPOINT` env support for story GraphQL separation
- Added `getStoryClient()` and `fetchStoryGQLData()` for story-specific GQL usage

## Removals  
- Removed all direct `fetch(URL_STATIC_*)` usages for JSON reading in favor of `readStaticJson()`
- Removed unused `param` for editor-choice fetch
- Removed fallback error-swallowing logic in JSON fetches that redundantly returned `{ postsData: [], jsonPostsCount: 0 }`

## Changes  
- Updated multiple action files (`app/actions.ts`, `app/actions-general.ts`, etc.) to use `readStaticJson()` instead of `fetch`
- Reorganized `config.ts`:
  - Added `STATIC_JSON_*` constants for relative JSON paths
  - Kept `URL_STATIC_*` for fallback HTTP URLs (with `TIMESTAMP_FOR_CACHE`)
- Refactored GraphQL logic:
  - Replaced `fetchGQLData` with `fetchStoryGQLData` in `story/` and `external/` pages
  - Updated `apollo-client.ts` to support `getStoryClient()` with `API_STORY_GQL_ENDPOINT`

## Testing  
- Verified fallback fetch behavior by simulating fs read failure
- Confirmed schema validation still works after replacing fetch logic

## Screenshots  
N/A

## Todos  
- Verify performance improvement from fs-first strategy  
- Ensure Cloud Build has access to GCS_FUSE_MOUNT_DIR in all environments  

## Task Links  
[前端改打新 end point - Asana](https://app.asana.com/1/614399484723017/project/1212507937806015/task/1212562795886200?focus=true)

Commit:
feat: add static story GraphQL endpoint for story and external pages

- Add API_STATIC_STORY_ENDPOINT config for static-mounted GraphQL API
- Create getStoryClient() for story-specific Apollo Client instance
- Add fetchStoryGQLData() to query data from static endpoint
- Update story and external pages to use new static endpoint
- Keep original fetchGQLData() intact for other pages

Commit:
feat(static-json): implement fs-first static JSON loading with unified timestamped fetch fallback

- Add readStaticJson() utility that prioritizes reading JSON from the GCS FUSE mounted filesystem (/statics)
- Implement fallback to HTTPS fetch when fs access fails
- Apply timestamp query parameter to all fallback fetch requests to prevent CDN/browser caching
- Centralize static JSON path definitions in config.ts
- Remove outdated URL_STATIC_* patterns in favor of unified loader logic
- Support local development through configurable GCS_FUSE_MOUNT_DIR

Commit:
fix(topic): limit transformed topic list to a maximum of six items

Commit:
chore(static-json): add 'use server' directive to read-static-json utility

Rice9547 commented on utils/read-static-json.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  // Fallback to HTTP fetch
+  const fallbackUrl = `https://${bucketDomain}/json/${cleanPath}?t=${Date.now() / 100}`
```
這個寫法應該會因為 cache 失效造成沒辦法命中 CDN？

fzoonoise replied:
原先是原本的常數網址都設定為加入時間戳記，想說維持一樣的寫法，已更新，應該可以避免你說的問題吧
https://github.com/mirrordaily/mirror-daily-web/pull/619/commits/a05ababb644766d6aff6201856ec31de92b186c9

Rice9547 replied:
不是 no-cached，這樣會造成推版後被 browser 卡住的問題
是你上個 commit 的寫法在每次 fetch 都會換一個 timestamp，但原本的寫法其實是第一次掃到這個檔案的時候就會設定好 timestamp (singleton)，來實現每次 re-build 都不會被 browser cache

fzoonoise replied:
原來如此，感謝你費心解釋，學到一招，已再更新於以下
https://github.com/mirrordaily/mirror-daily-web/pull/619/commits/4903d67c259db2487aee9a3e50d09a7de8aa0848

fzoonoise replied:
額外修正掉時間戳重複的問題(多的 log 已經在後面 commit 移掉了)
https://github.com/mirrordaily/mirror-daily-web/pull/619/commits/0e4507edef7bb5e0ec9cd09efa27be059fe59750

Commit:
refactor(static-json): use `cache: 'no-store'` for fetch fallback

Commit:
fix: restore build-level cache busting for static JSON fallback

nihclil commented on constants/config.ts:
```diff
@@ -73,40 +73,60 @@ switch (ENV) {
+const URL_STATIC_CREATIVITY_SHORTPAGE = `${JSON_FILE_PATH}/${STATIC_JSON_CREATIVITY_SHORTPAGE}`
 // section page
-const URL_STATIC_SECTION_NEWS = `${JSON_FILE_PATH}/latest/latest_content_section`
+const STATIC_JSON_SECTION_NEWS = 'latest/latest_content_section'
```
這種有分頁的，頁面會壞掉

fzoonoise replied:
已先修正你指出 header 資料來源有問題，修正如下：
https://github.com/mirrordaily/mirror-daily-web/pull/619/commits/0e4507edef7bb5e0ec9cd09efa27be059fe59750
https://github.com/mirrordaily/mirror-daily-web/pull/619/commits/b7f056db4a8198db1516591a0c7efa512878b388

Commit:
fix(data-fetching): adjust timestamp constants and fix header json parsing issue

Commit:
fix(data-fetching): resolve Zod errors from readStaticJson usage

This commit removes the unnecessary 'z.promise()' wrappers from all affected server actions in 'app/actions.ts' and 'app/actions-general.ts', and correct the Zod parsing logic to directly validate the resolved JSON objects.

Commit:
Merge remote-tracking branch 'upstream/dev' into feat/2025-12-24_update-api-endpoint

# Conflicts:
#	app/actions.ts
#	constants/config.ts

Rice9547 reviewed (approved):
LGTM

hcchien commented on app/actions.ts:
```diff
@@ -206,31 +208,28 @@ const transformEditorChoices = (
 export const fetchEditorChoices = async (): Promise<
   ParameterOfComponent<typeof EditorChoiceMain>
 > => {
-  const param = String(Date.now()).slice(0, 8)
```
這個好像可以留著，不然會常常遇到更新太慢的問題

fzoonoise replied:
我檢查過目前 server side 需要走 fallback fetch 的使用情境，發現本來就都有做類似的 cache invalidation 的處理，後續已將這段邏輯統一放到共用的 [utils/read-static-json.ts ](https://github.com/mirrordaily/mirror-daily-web/pull/619/files/86c78a9032a0d1994fdabb71de26a362667d130a#diff-dd9db9ed5ad98024c969ef2659498867eea40209d6770036b71f748311614b15)。目前 cache busting 會在該共用 function 中以 build-time timestamp 處理，而不再需要在各個 action 裡保留每次 request 產生的 Date.now() 參數。

hcchien reviewed (approved):
LGTM
